File: Tool/PreUtil/src/output/javascript/medium/MaximumStarSumOfAGraph.js
/**https://leetcode.com/problems/maximum-star-sum-of-a-graph/ */
//There is an undirected graph consisting of n nodes numbered from 0 to n - 1. You are given a 0-indexed integer array vals of length n where vals[i] denotes the value of the ith node.
//You are also given a 2D integer array edges where edges[i] = [ai, bi] denotes that there exists an undirected edge connecting nodes ai and bi.
//A star graph is a subgraph of the given graph having a center node containing 0 or more neighbors. In other words, it is a subset of edges of the given graph such that there exists a common node for all edges.
//The image below shows star graphs with 3 and 4 neighbors respectively, centered at the blue node.


/**
 * @param {number[]} vals
 * @param {number[][]} edges
 * @param {number} k
 * @return {number}
 */
var maxStarSum = function(vals, edges, k) {
    let n = vals.length;
    let result = 0;
    let graph = new Array(n).fill(0).map(() => new Array());
    for(let i = 0; i < edges.length; i++){
        let [a, b] = edges[i];
        graph[a].push(b);
        graph[b].push(a);
    }
    let dp = new Array(n).fill(0).map(() => new Array(k + 1).fill(0));
    let dfs = (node, parent) => {
        for(let i = 0; i < graph[node].length; i++){
            let child = graph[node][i];
            if(child != parent){
                dfs(child, node);
                for(let j = k; j >= 1; j--){
                    for(let l = 0; l < j; l++){
                        dp[node][j] = Math.max(dp[node][j], dp[node][j - l - 1] + dp[child][l]);
                    }
                }
            }
        }
        for(let i = k; i >= 1; i--){
            dp[node][i] = dp[node][i - 1] + vals[node];
        }
        dp[node][0] = 0;
    }
    dfs(0, -1);
    return dp[0][k];
    
};

    

File: Tool/PreUtil/src/output/c/medium/MaximumStarSumOfAGraph.c
/**https://leetcode.com/problems/maximum-star-sum-of-a-graph/ */
//There is an undirected graph consisting of n nodes numbered from 0 to n - 1. You are given a 0-indexed integer array vals of length n where vals[i] denotes the value of the ith node.
//You are also given a 2D integer array edges where edges[i] = [ai, bi] denotes that there exists an undirected edge connecting nodes ai and bi.
//A star graph is a subgraph of the given graph having a center node containing 0 or more neighbors. In other words, it is a subset of edges of the given graph such that there exists a common node for all edges.
//The image below shows star graphs with 3 and 4 neighbors respectively, centered at the blue node.


int maxStarSum(int* vals, int valsSize, int** edges, int edgesSize, int* edgesColSize, int k){
    int result = 0;
    int dp[valsSize];
    for(int i = 0; i < valsSize; i++){
        dp[i] = 0;
    }
    for(int i = 1; i < valsSize; i++){
        for(int j = 0; j < i; j++){
            if(vals[i] == vals[j] * vals[j]){
                dp[i] = dp[j] + 1;
                result = result > dp[i] ? result : dp[i];
            }
        }
    }
    return result == 0 ? -1 : result + 1;

}

    